### ISSUE TITLE
No job functions found, using docker and model v2

### ISSUE THREAD

**Issue by martin-loland-4ss** [Azure/azure-functions-host#9458]
No job functions found running a model 2 functions app in container.

```
func --version
4.0.5198
```
#### Repro steps

```
func init --worker-runtime python -m v2 --docker
func new -l python -t "HTTP trigger"
docker build -t func:latest .
docker run -p 80:80 -it func:latest
```

**Console shows following**
```
(...)
info: Host.Startup[327]
      1 functions found
info: Host.Startup[315]
      0 functions loaded
info: Host.Startup[0]
      Generating 0 job function(s)
warn: Host.Startup[0]
      No job functions found. Try making your job classes and methods public. If you're using binding extensions (e.g. Azure Storage, ServiceBus, Timers, etc.) make sure 
you've called the registration method for the extension(s) in your startup code (e.g. builder.AddAzureStorage(), builder.AddServiceBus(), builder.AddTimers(), etc.).     
info: Microsoft.Azure.WebJobs.Script.WebHost.WebScriptHostHttpRoutesManager[0]
      Initializing function HTTP routes
      No HTTP routes mapped
(...)
Hosting environment: Production
Content root path: /azure-functions-host
Now listening on: http://[::]:80
Application started. Press Ctrl+C to shut down.
```
#### Expected behavior

That functions work in the container. It works as expected when running using `func start`.

#### Content of files

*Dockerfile*
```
# To enable ssh & remote debugging on app service change the base image to the one below
# FROM mcr.microsoft.com/azure-functions/python:4-python3.10-appservice
FROM mcr.microsoft.com/azure-functions/python:4-python3.10

ENV AzureWebJobsScriptRoot=/home/site/wwwroot \
    AzureFunctionsJobHost__Logging__Console__IsEnabled=true

COPY requirements.txt /
RUN pip install -r /requirements.txt

COPY . /home/site/wwwroot
```

*host.json*
```json
{
  "version": "2.0",
  "logging": {
    "applicationInsights": {
      "samplingSettings": {
        "isEnabled": true,
        "excludedTypes": "Request"
      }
    }
  },
  "extensionBundle": {
    "id": "Microsoft.Azure.Functions.ExtensionBundle",
    "version": "[4.*, 5.0.0)"
  }
}
```

*local.settings.json*
```json
{
  "IsEncrypted": false,
  "Values": {
    "FUNCTIONS_WORKER_RUNTIME": "python",
    "AzureWebJobsFeatureFlags": "EnableWorkerIndexing",
    "AzureWebJobsStorage": "UseDevelopmentStorage=true"
  }
}
```

*function_app.py*
```python
import azure.functions as func
import logging

app = func.FunctionApp()


@app.route(route="HttpTrigger", auth_level=func.AuthLevel.ANONYMOUS)
def HttpTrigger(req: func.HttpRequest) -> func.HttpResponse:
    logging.info('Python HTTP trigger function processed a request.')

    name = req.params.get('name')
    if not name:
        try:
            req_body = req.get_json()
        except ValueError:
            pass
        else:
            name = req_body.get('name')

    if name:
        return func.HttpResponse(f"Hello, {name}. This HTTP triggered function executed successfully.")
    else:
        return func.HttpResponse(
             "This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response.",
             status_code=200
        )
```

**Comment by martin-loland-4ss** [comment]
Got it working by adding two env vars in the Dockerfile:

```
ENV AzureWebJobsScriptRoot=/home/site/wwwroot \
    AzureFunctionsJobHost__Logging__Console__IsEnabled=true \
    AzureWebJobsFeatureFlags=EnableWorkerIndexing \
    AzureWebJobsStorage=UseDevelopmentStorage=true
```
Ref: https://stackoverflow.com/questions/75519932/azure-function-python-model-2-in-docker-container

**Comment by bhagyshricompany** [comment]
thanks for informing. Hope soln is working now.

**Comment by grobf5** [comment]
# Reproducing — Python v2 model, both Premium V3 and Flex Consumption

Confirming this on the v2 programming model with `AzureWebJobsFeatureFlags=EnableWorkerIndexing` correctly set. Host reports "No job functions found" — same symptom as "0 functions found (Custom)" in [#9720](https://github.com/Azure/azure-functions-host/issues/9720).

- Reproduced on Linux Premium V3 (`P1v3`) and Flex Consumption (`FC1`).
- `function_app.py` has `@app.route`, `@app.timer_trigger`, and `@app.warmup_trigger` decorators.
- Tried every deploy mechanism MS documents: `func publish --build remote`, `az functionapp deployment source config-zip`, `azurerm_linux_function_app.zip_deploy_file`, Flex Consumption's blob-container deploy per [Azure-Samples canonical TF sample](https://github.com/Azure-Samples/azure-functions-flex-consumption-samples/tree/main/IaC/terraformazurerm). All four land the deploy artifact correctly, then host reports 0 functions.

Full environment + a public reproducing repo at [Azure/azure-functions-python-worker#1338](https://github.com/Azure/azure-functions-python-worker/issues/1338).

This affects production Terraform deployments on both Microsoft-recommended Python v2 plans. Any visibility into the root cause / workaround would unblock real customer workloads — happy to provide additional diagnostics.

**Comment by grobf5** [comment]
**Resolved on our side.** Posting the root cause + fix in case it helps anyone else hitting "0 functions found (Custom)" / `admin/functions` returns `[]` / every `/api/*` route 404 on the Python v2 model.

There were **two independent bugs** stacked on top of each other producing the same symptom:

### Bug 1 — silent `AttributeError` in the v2 worker indexer (this one is, IMO, still a real worker bug)

Our `function_app.py` declared the warmup trigger as:

    @app.warmup_trigger()
    def warmup(context: func.Context) -> None:
        ...

The correct decorator name in `azure-functions>=1.20` is `warm_up_trigger` (underscore between `warm` and `up`) and it takes the trigger arg name as a positional argument:

    @app.warm_up_trigger("warmup")
    def warmup(context: func.Context) -> None:
        ...

Calling the wrong name raises `AttributeError: 'FunctionRegister' object has no attribute 'warmup_trigger'` *inside the worker's indexer when it imports `function_app.py`*. The worker swallows that exception entirely:

- No traceback in any log stream we could find (Kudu, Log Stream, App Insights `traces`/`exceptions`, host stdout)
- Host log shows the standard `Reading functions metadata (Custom) → 0 functions found (Custom)` — same line you'd see for an empty file
- `admin/host/status` returns `"state": "Running"` — the host is healthy, the worker just produced an empty function list and the host accepted it

**This is the bug worth keeping these issues open for.** A typo in *any one* decorator name silently turns the entire app into a phantom — *all* decorated functions in the file vanish, not just the broken one, and you get no diagnostic anywhere. From the host's perspective the worker indexed successfully and reported zero functions. There's no way for a developer to distinguish "I have an import-time exception in my indexed module" from "I correctly have zero functions" without external instrumentation. A surfaced `WorkerIndexingFailed` event (or even a `traces` row in App Insights with the AttributeError) would have saved us roughly a week.

### Bug 2 — Flex Consumption does *not* auto-mount a zip uploaded to the `deploymentpackage` container

Our Terraform was uploading `function-app.zip` directly to the function app's Flex storage container via `azurerm_storage_blob`, on the assumption (drawn from the [Azure-Samples canonical TF sample](https://github.com/Azure-Samples/azure-functions-flex-consumption-samples/tree/main/IaC/terraformazurerm)) that the Flex runtime mounts the latest zip in that container automatically. **It does not.** Flex only mounts code pushed via OneDeploy — the `az functionapp deployment source config-zip` path, which hits `/api/deploy/*`, uploads the zip, *and* runs the sync-triggers handshake that registers functions with the host.

Even with Bug 1 fixed, a direct blob upload to `deploymentpackage` produces an empty `wwwroot` and the same "0 functions found (Custom)" symptom — because the host never sync-triggers and never knows about the uploaded artifact. Switching to a `null_resource` running `az functionapp deployment source config-zip` post-create resolved this layer.

(The MS canonical TF sample is, as best I can tell, accurate but underspecified — it shows the storage container being created and uploaded to, but the actual code-push step is glossed as "deploy your function code" with no mention that direct blob upload is insufficient on Flex.)

### Bug 3 (sort of) — Flex Consumption has no Kudu log stream, so Bug 1 is invisible without App Insights

Flex Consumption has no SCM site / Kudu log stream / Log Stream blade. Without [workspace-based] Application Insights wired in via `APPLICATIONINSIGHTS_CONNECTION_STRING`, you have **zero** visibility into worker startup, indexer behavior, or any exception path. We spent days chasing this on Flex without AI before realizing it. (Even with AI, Bug 1's swallowed AttributeError still didn't surface — but at least the rest of the stack became visible.)

### Net fix

The combination that works on Flex Consumption (`FC1`, Python 3.12, v2 model):

1. Audit every decorator in `function_app.py` against the current `azure-functions` SDK — typos vanish silently, with no diagnostic anywhere.
2. Push code via `az functionapp deployment source config-zip`, not via direct blob upload to `deploymentpackage`.
3. Wire workspace-based App Insights from day one — set `APPLICATIONINSIGHTS_CONNECTION_STRING` in `app_settings`. Without it, the Flex runtime is a black box.
4. Keep the rest of the recipe: `AzureWebJobsFeatureFlags=EnableWorkerIndexing`, UAMI with Blob Owner + Queue Contributor + Table Contributor on the function-runtime SA, `cryptography>=43,<45` (the `>=46` ABI break is a separate issue).

Same combination works on Linux Premium V3 with `func azure functionapp publish --python --build remote`, with the same caveat about the silent decorator typo.

### Ask for the maintainers

The Flex auto-mount and AI-required points are documentation gaps — easy fixes on the docs side.

Bug 1 (silent indexer `AttributeError`) is a real worker behavior worth fixing. At minimum: a `WorkerIndexingFailed` host log line with the exception type, or an App Insights `exceptions` row when `function_app.py` raises during import. Happy to repro for triage if useful — the minimal case is a single decorator name typo against the current SDK on a v2 app, no other code or deps required.
